File: algorithms/python/findRightInterval/findRightInterval.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/find-right-interval/
# Author: Miao Zhang
# Date:   2021-02-11

class Solution:
    def findRightInterval(self, intervals: List[List[int]]) -> List[int]:
        import bisect
        itvs = sorted((it[0], i) for i, it in enumerate(intervals))
        res = []
        for itv in intervals:
            idx = bisect.bisect_left(itvs, (itv[1],))
            res.append(itvs[idx][1] if idx < len(intervals) else -1)
        return res

File: algorithms/cpp/findRightInterval/findRightInterval.cpp
// Source: https://leetcode.com/problems/find-right-interval/
// Author: Miao Zhang
// Date:   2021-02-11

class Solution {
public:
    vector<int> findRightInterval(vector<vector<int>>& intervals) {
        vector<int> res, starts;
        unordered_map<int, int> m;
        for (int i = 0; i < intervals.size(); i++) {
            m[intervals[i][0]] = i;
            starts.push_back(intervals[i][0]);
        }
        sort(starts.rbegin(), starts.rend());
        for (auto it: intervals) {
            int i = 0;
            for (; i < starts.size(); i++) {
                if (it[1] > starts[i]) break;
            }
            res.push_back((i > 0) ? m[starts[i - 1]] : -1);
        }
        return res;
    }
};

/***********************************************************
***********************************************************/
class Solution {
public:
    vector<int> findRightInterval(vector<vector<int>>& intervals) {
        vector<int> res;
        map<int, int> m;
        for (int i = 0; i < intervals.size(); i++) {
            m[intervals[i][0]] = i;
        }
        for (auto itv: intervals) {
            auto it = m.lower_bound(itv[1]);
            if (it == m.end()) {
                res.push_back(-1);
            } else {
                res.push_back(it->second);
            }
        }
        return res;
    }
};
